app/views/documents/show/_history_tab.html.erb

```diff
@@ -1,31 +1,39 @@
 <% displayed_edition_number = nil %>
 
-<% @document.timeline_entries.includes(:user).order(created_at: :desc).includes(:retirement).includes(:removal).each do |entry| %>
-  <div class="app-timeline-entry govuk-!-margin-bottom-6">
-    <% if displayed_edition_number != entry.edition_number %>
-      <h3 class="govuk-heading-m"><%= t "documents.history.edition_header", number: entry.edition_number.ordinalize %></h3>
-      <% displayed_edition_number = entry.edition_number %>
-    <% end %>
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <% @document.timeline_entries.includes(:user).order(created_at: :desc).includes(:retirement, :removal, :internal_note).each do |entry| %>
+      <div class="app-timeline-entry govuk-!-margin-bottom-6">
+        <% if displayed_edition_number != entry.edition_number %>
+          <h3 class="govuk-heading-m"><%= t "documents.history.edition_header", number: entry.edition_number.ordinalize %></h3>
+          <% displayed_edition_number = entry.edition_number %>
+        <% end %>
 
-    <h4 class="govuk-heading-s govuk-!-margin-bottom-1">
-      <%= t "documents.history.entry_types.#{entry.entry_type}" %>
-    </h4>
+        <h4 class="govuk-heading-s govuk-!-margin-bottom-1">
+          <%= t "documents.history.entry_types.#{entry.entry_type}" %>
+        </h4>
 
-    <% if entry.retirement.present? %>
-      <%= entry.retirement.explanatory_note %>
-    <% elsif entry.removal.present? %>
-      <% if entry.removal.explanatory_note.present? %>
-        <p><%= entry.removal.explanatory_note %></p>
-      <% end %>
-      <% if entry.removal.redirect %>
-        <p>Redirected to <%= link_to(Plek.new.website_root + entry.removal.alternative_path) %></p>
-      <% elsif entry.removal.alternative_path.present? %>
-        <p><%= link_to(Plek.new.website_root + entry.removal.alternative_path) %></p>
-      <% end %>
-    <% end %>
+        <div class="app-timeline-entry__dateline">
+          <%= entry.created_at.strftime("%d %B %Y at %l:%M%P") %> by <%= entry.username_or_unknown %>
+        </div>
 
-    <div class="app-timeline-entry__dateline">
-      <%= entry.created_at.strftime("%d %B %Y at %l:%M%P") %> by <%= entry.username_or_unknown %>
-    </div>
+        <% if entry.internal_note.present? %>
+          <p><%= sanitize(simple_format(entry.internal_note.body).gsub(URI.regexp, '<a href="\0">\0</a>')) %></p>
```

Comment:
I think there's a couple of problems with this and I'm also a bit confused.

So reading the docs [simple_format](https://apidock.com/rails/ActionView/Helpers/TextHelper/simple_format) will convert into a <p> element so I think this means that this doesn't need wrapping inside a p element? this also sanitizes right?

Then we do the gsub on the resulting HTML - which is a nice approach I didn't consider you could do that and thought you might need https://github.com/vmg/rinku or https://github.com/tenderlove/rails_autolink - though I guess that means we lose mailto links which would be nice to have. It'd also be good to just restrict the URI links to http and https which I think you can do with `URI.regexp(%w[http https])`

Then where I get confused is the sanitize call as I'd have thought that would have stripped out the elements we just added? but I assume that was fine with you trying it out?

Then finally I'd have thought (unless rails added it along the way somewhere) that this needs a html_safe call on it.

It might be easier to create a view helper method that we can test `simple_format_with_links` that takes simple_format arguments and any needed for links. What do you think?

Reply:
You're correct in that `simple_format` wraps the text in a `<p>` tag, so the `<p>` i've added isn't needed which i will remove and yep we could restrict URI links as you've stated to http and https . 

The sanitize method works in a similar way to `html_safe` in terms of rendering the actual html, however it will strip the text of any potentially unsafe tags leaving the href links intact. As just calling `html_safe` on the text would create a vulnerability where publishers (highly unlikely i know) could submit a script which would be executed when the `internal_note` was rendered. This is a snippet from the docs about the sanitize method.

```Sanitizes HTML input, stripping all but known-safe tags and attributes.

It also strips href/src attributes with unsafe protocols like javascript:, while also protecting against attempts to use Unicode, ASCII, and hex character references to work around these protocol filters. All special characters will be escaped.

The default sanitizer is Rails::Html::WhiteListSanitizer. See Rails HTML Sanitizers for more information.```

I tested it locally by submitting links and hyperlinks were displayed on the internal note with the correct line breaks. However, if we were to keep this approach i'd agree that it'd probably be best to move it to its own method so its testable.

I could write another regexp that swaps any email address with it's corresponding `mailto` link, however this maybe feels like a case of recreating the wheel?

So, in summary i'd say if we want the mailto links we could use one of the gems you suggested instead of my approach but i'll leave the decision up to you :)

Reply:
Yeah if gem seems the best way to include mailto then give that a whirl.

I'm a bit confused how the santize method knows to allow `<a>`, `<p>` and `<br>` - are they considered the "known-safe tags" 

I thought that after we ran simple_format we'd then have a string that's safe to run html safe on as html entities would have been encoded accordingly. 

Probably good to have this as a tested method since it's generated quite a lot of conversation with us that could all be represented in the test. Sound good to you?